### DISCUSSION TITLE
Treatment of templated classes: Generic vs Explicit type specification

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#21033] [Q&A General] [answered]
Dear Moose experts,
As part of my recent developments on `ArrayAuxKernels` i am trying to write the list of AuxKernel values on to the auxvariables through the `setNodalValue` function. One common thing i observed while implementing this function is that the templated class functions are executed both for the case of explict specification of the desired type `(RealEigenVector)` and for the generic type `(OutputType)`. As a result i could not confine my developments to a single function relevant to my type. The following example will make this case clear.
Case 1:
```
template <>
void
MooseVariableData<RealEigenVector>::computeValues()
{
```
case 2:
```
template <typename OutputType>
void
MooseVariableData<OutputType>::computeValues()
{
```
I expect any calls pertaining to `ArrayAuxKernel` would go to case 1 rather than case 2. Execution of both these cases can sometimes lead to overwrite of data set by each of these functions and is generally not advisable. My question is how to restrict calls to only case 1 where the type is specified explicity inside <>.

**Comment by abarun22** [comment]
Another related question is that the following code would not compile throwing the below error. 
```
template <>
void
MooseVariableData<RealEigenVector>::setNodalValue(const RealEigenVector & value, unsigned int idx)
{
      _has_dof_values = true;
      _nodal_value = value;
      for (unsigned int i = 0; i < _u.size(); i++)
       _u[i] = value(i);
}
```
Error:
```
/home/abalasub/Working/Projects/moose/framework/src/variables/MooseVariableData.C: In member function ‘void MooseVariableData<OutputType>::setNodalValue(const OutputType&, unsigned int) [with OutputType = Eigen::Matrix<double, -1, 1>]’:
/home/abalasub/Working/Projects/moose/framework/src/variables/MooseVariableData.C:1620:26: error: no match for ‘operator=’ (operand types are ‘Eigen::Matrix<double, -1, 1>’ and ‘const double’)
         _u[qp] = value(qp);
```
The variable _u is defined in `MooseVariableData.h` as follows
`FieldVariableValue _u;`

As can be seen from the compile log the type of RHS variable is evalated wrongly and hence the error. I am sure the explicit declaration of the type of `value` will certainly resolve this to `Eigen::Matrix<double, -1, 1>`, but in this case it only sees a `const double`. The following is the prototype declation in `MooseVariableData.h`.

```
template <typename OutputType>
class MooseVariableData
{
void setNodalValue(const OutputType & value, unsigned int idx = 0);
```
Any idea what's going on here?
Kind regards,
Arun

**Comment by GiudGiud** [comment, marked as answer]
Hello

I think you want to instantiate this template to ~~RealVectorValue~~ RealEigenVector
So OutputType becomes ~~RealVectorValue~~ RealEigenVector. 

See GenericConstantMaterial for example. 
The whole class is defined with the parameter is_ad, then it's instantiated to true and false. 

We want to do the same here, but with OutputType being replaced by ~~RealVectorValue~~ RealEigenVector.

What's going on with your error is that in
       _u[i] = value(i);
the quantites on both sides have different types.
value(i) is a single value, because value is a vector
_u is a tensor material property [i] I believe should be [_qp](i) (indexing at quadrature point then into a row/column (I have not checked) )

Please look at the documentation for Eigen::Matrix to determine the appropriate indexing into _u[_qp]

Guillaume

**Reply by abarun22** [reply]
Hi Guillaume,
I think RealVectorValue is the relevant type to be used for this class template. On the question of preferrential execution, i wanted to ask if the templated class is not instantiated for a particular type, but a function has been defined for this type, will that function be executed? Consider the folowing case.

```
template <>
void
MooseVariableData<RealEigenVector>::setNodalValue(const RealEigenVector & value, unsigned int idx)
{
```

And i comment out this line in the end:

```//template class MooseVariableData<RealEigenVector>;```

When i checked this, the function is still executed, which seems quiet weird, as only the explicitly instantiated classes will be called.

Kind regards,
Arun

**Comment by abarun22** [comment]
I am working towards replacing the `RealEigenVector` type with `RealVectorValue` in all of my array auxkernel developments. After doing this  migration, i encountered this weird error when it reaches the last line in the below code
```
ArrayRankTwoAux.C:46:9: error: ‘libMesh::RealVectorValue {aka class libMesh::VectorValue<double>}’ has no member named ‘resize’; did you mean ‘rebind’?
   evalm.resize(teval*_nsubvar);
```

```
template <bool is_ad>
RealVectorValue
ArrayRankTwoAuxTempl<is_ad>::computeValue()
{
  auto nqpts=this->_qrule->n_points();
  RealVectorValue  evalm;
  std::vector<std::pair<unsigned int, unsigned int>> tindices = {{0,0},{1,1},{2,2},{0,1},{0,2},{1,2}};
  auto teval = nqpts*tindices.size();
  evalm.resize(teval*_nsubvar);
```
Somehow the compiler is interpreting this type wrongly and look for `resize` inside a class that is mapped to RealVectorValue. Any suggestions/leads here?
Kind regards,
Arun

**Reply by GiudGiud** [reply]
@abarun22 
No I was wrong, it's RealEigenVector you want not RealVectorValue. The latter is fixed 3 size

**Reply by GiudGiud** [reply]
I was trying to make a point about the template instantiation, did not think I was mentioning the wrong type
